File: LeetCode/1691. Maximum Height by Stacking Cuboids/solution.py
class Solution:
    def maxHeight(self, A):
        A = [[0, 0, 0]] + sorted(map(sorted, A))
        dp = [0] * len(A)
        for j in range(1, len(A)):
            for i in range(j):
                # 充分必要条件！
                if all(A[i][k] <= A[j][k] for k in range(3)):
                    dp[j] = max(dp[j], dp[i] + A[j][2])
        return max(dp)        



from functools import lru_cache
class Solution:
    def maxHeight(self, cuboids: List[List[int]]) -> int:
        cuboids.append([0,0,0])
        cuboids.sort(key=lambda x: x[0]*x[1]*x[2])
        n = len(cuboids)
        dp = [[0]*3 for _ in range(n)]
        res = 0
        for i in range(1,n):
            for pi in range(3):
                cur = [cuboids[i][k] for k in range(3) if k != pi]
                for j in range(i):
                    for pj in range(3):
                        if cuboids[j][pj] > cuboids[i][pi]:
                            continue
                        pre = [cuboids[j][k] for k in range(3) if k != pj]
                        if all(pre[i] <= cur[i] for i in range(2)) or all(pre[i] <= cur[1-i] for i in range(2)):
                            dp[i][pi] = max(dp[i][pi], cuboids[i][pi]+dp[j][pj])
                            res = max(res, dp[i][pi])

        # def dfs(idx,pos):
        #     res = 0
        #     cur = [cuboids[idx][k] for k in range(3) if k != pos]
        #     for j in range(idx):
        #         for p in range(3):
        #             pre = [cuboids[j][k] for k in range(3) if k != p]
        #             if all(pre[i] <= cur[i] for i in range(2)) or all(pre[i] <= cur[1-i] for i in range(2)):
        #                 res = max(res, cuboids[idx][pos]+dfs(j,p))
        # print(cuboids)
        # print(dp)
        # print(max(dp))
        return res

File: LeetCode/1691. Maximum Height by Stacking Cuboids/solution.cpp
class Solution {
public:
    int maxHeight(vector<vector<int>>& cuboids) {
        int n = cuboids.size();
        vector<tuple<int, int, int, int>> v;
        for (int i = 0; i < n; ++i) {
            const auto& cubic = cuboids[i];
            v.emplace_back(cubic[0], cubic[1], cubic[2], i);
            v.emplace_back(cubic[0], cubic[2], cubic[1], i);
            v.emplace_back(cubic[1], cubic[0], cubic[2], i);
            v.emplace_back(cubic[1], cubic[2], cubic[0], i);
            v.emplace_back(cubic[2], cubic[0], cubic[1], i);
            v.emplace_back(cubic[2], cubic[1], cubic[0], i);
        }
        
        sort(v.begin(), v.end());
        
        vector<int> f(n * 6);
        for (int i = 0; i < n * 6; ++i) {
            auto [wi, li, hi, idi] = v[i];
            for (int j = 0; j < i; ++j) {
                auto [wj, lj, hj, idj] = v[j];
                if (wj <= wi && lj <= li && hj <= hi && idj != idi) {
                    f[i] = max(f[i], f[j]);
                }
            }
            f[i] += hi;
        }
        
        return *max_element(f.begin(), f.end());
    }
};


// 作者：zerotrac2
// 链接：https://leetcode-cn.com/problems/maximum-height-by-stacking-cuboids/solution/dui-die-chang-fang-ti-de-zui-da-gao-du-b-qzgy/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

// class Solution {
// public:
//     int maxHeight(vector<vector<int>>& cuboids) {
//         int n = cuboids.size();
//         for (auto& cubic: cuboids) {
//             sort(cubic.begin(), cubic.end());
//         }
        
//         // 保证枚举关系拓扑性的排序都可以
//         // sort(cuboids.begin(), cuboids.end());
//         sort(cuboids.begin(), cuboids.end(), [](const auto& u, const auto& v) {
//             return u[2] + u[0] + u[1] < v[2] + v[0] + v[1];
//         });
        
//         vector<int> f(n);
//         for (int i = 0; i < n; ++i) {
//             for (int j = 0; j < i; ++j) {
//                 if (cuboids[j][0] <= cuboids[i][0] && cuboids[j][1] <= cuboids[i][1] && cuboids[j][2] <= cuboids[i][2]) {
//                     f[i] = max(f[i], f[j]);
//                 }
//             }
//             f[i] += cuboids[i][2];
//         }
        
//         return *max_element(f.begin(), f.end());
//     }
// };


// 作者：zerotrac2
// 链接：https://leetcode-cn.com/problems/maximum-height-by-stacking-cuboids/solution/dui-die-chang-fang-ti-de-zui-da-gao-du-b-qzgy/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。